File: src/main/java/com/kingdom/model/IncompleteCard.java
package com.kingdom.model;

import java.util.LinkedList;
import java.util.Queue;

/**
 * Created by IntelliJ IDEA.
 * User: John
 * Date: 3/31/11
 * Time: 6:37 AM
 */
public abstract class IncompleteCard {

    protected String cardName;
    protected Game game;
    protected Queue<CardAction> extraCardActions = new LinkedList<CardAction>();
    protected boolean endTurn;
    protected boolean allActionsSet;

    protected IncompleteCard(String cardName, Game game) {
        this.cardName = cardName;
        this.game = game;
        game.setIncompleteCard(this);
    }

    public abstract boolean isCompleted();

    public Queue<CardAction> getExtraCardActions() {
        return extraCardActions;
    }

    public abstract void setWaitingDialogs();

    public String getCardName() {
        return cardName;
    }

    public abstract void setPlayerActionCompleted(int playerId);

    public boolean isEndTurn() {
        return endTurn;
    }

    public void setEndTurn(boolean endTurn) {
        this.endTurn = endTurn;
    }

    public Game getGame() {
        return game;
    }

    public synchronized void actionFinished(Player player) {
        if (!player.isShowCardAction()) {
            setPlayerActionCompleted(player.getUserId());
            if (!extraCardActions.isEmpty()) {
                game.setPlayerCardAction(game.getCurrentPlayer(), extraCardActions.remove());
            }
        }
        setWaitingDialogs();
    }

    public abstract void allActionsSet();
}

File: src/main/kotlin/com/kingdom/model/IncompleteCard.kt
package com.kingdom.model

import java.util.LinkedList
import java.util.Queue

abstract class IncompleteCard protected constructor(val cardName: String, val game: OldGame) {

    var extraOldCardActions: Queue<OldCardAction> = LinkedList()
        protected set

    var isEndTurn: Boolean = false

    protected open var isAllActionsSet: Boolean = false

    abstract val isCompleted: Boolean

    init {
        game.incompleteCard = this
    }

    abstract fun setWaitingDialogs()

    abstract fun setPlayerActionCompleted(playerId: Int)

    @Synchronized
    fun actionFinished(player: OldPlayer) {
        if (!player.isShowCardAction) {
            setPlayerActionCompleted(player.userId)
            if (!extraOldCardActions.isEmpty()) {
                game.setPlayerCardAction(game.currentPlayer!!, extraOldCardActions.remove())
            }
        }
        setWaitingDialogs()
    }

    abstract fun allActionsSet()
}
